// @applitools/eyes-sdk-core/lib/renderer/RenderRequest.js
'use strict'

const {ArgumentGuard} = require('@applitools/eyes-common')

/**
 * Encapsulates data required to start render using the RenderingGrid API.
 */
class RenderRequest {
  /**
   * @param {string} webhook
   * @param {string} url
   * @param {RGridDom} dom
   * @param {RGridResource[]} resources
   * @param {RenderInfo} [renderInfo]
   * @param {string} [platform]
   * @param {string} [browserName]
   * @param {Object} [scriptHooks]
   * @param {string[]} selectorsToFindRegionsFor
   * @param {boolean} sendDom
   * @param {string} renderId
   * @param {string} agentId
   */
  constructor({
    webhook,
    stitchingService,
    url,
    dom,
    resources,
    renderInfo,
    platform,
    browserName,
    scriptHooks,
    selectorsToFindRegionsFor,
    sendDom,
    renderId,
    agentId,
  } = {}) {
    ArgumentGuard.notNullOrEmpty(webhook, 'webhook')
    ArgumentGuard.notNull(url, 'url')
    ArgumentGuard.notNull(dom, 'dom')
    ArgumentGuard.notNull(resources, 'resources')

    this._webhook = webhook
    this._stitchingService = stitchingService
    this._url = url
    this._dom = dom
    this._resources = resources
    this._renderInfo = renderInfo
    this._platform = platform
    this._browserName = browserName
    this._renderId = renderId
    this._scriptHooks = scriptHooks
    this._selectorsToFindRegionsFor = selectorsToFindRegionsFor
    this._sendDom = sendDom
    this._agentId = agentId
  }

  /**
   * @return {string}
   */
  getWebhook() {
    return this._webhook
  }

  /**
   * @return {string}
   */
  getStitchingService() {
    return this._stitchingService
  }

  /**
   * @return {string}
   */
  getUrl() {
    return this._url
  }

  /**
   * @return {RGridDom}
   */
  getDom() {
    return this._dom
  }

  /**
   * @return {RGridResource[]}
   */
  getResources() {
    return this._resources
  }

  /**
   * @return {RenderInfo}
   */
  getRenderInfo() {
    return this._renderInfo
  }

  /**
   * @return {string}
   */
  getPlatform() {
    return this._platform
  }

  /**
   * @return {string}
   */
  getBrowserName() {
    return this._browserName
  }

  /**
   * @return {string}
   */
  getRenderId() {
    return this._renderId
  }

  /**
   * @return {string}
   */
  getAgentId() {
    return this._agentId
  }

  /**
   * @param {string} value
   */
  setAgentId(value) {
    this._agentId = value
  }

  /**
   * @param {string} value
   */
  setRenderId(value) {
    this._renderId = value
  }

  /**
   * @return {string}
   */
  getScriptHooks() {
    return this._scriptHooks
  }

  /**
   * @param {string} value
   */
  setScriptHooks(value) {
    this._scriptHooks = value
  }

  /**
   * @return {string[]}
   */
  getSelectorsToFindRegionsFor() {
    return this._selectorsToFindRegionsFor
  }

  /**
   * @param {string[]} value
   */
  setSelectorsToFindRegionsFor(value) {
    this._selectorsToFindRegionsFor = value
  }

  /**
   * @return {boolean}
   */
  getSendDom() {
    return this._sendDom
  }

  /**
   * @param {boolean} value
   */
  setSendDom(value) {
    this._sendDom = value
  }

  /**
   * @override
   */
  toJSON() {
    const resources = {}
    this.getResources().forEach(resource => {
      resources[resource.getUrl()] = resource.getHashAsObject()
    })

    const object = {
      webhook: this._webhook,
      stitchingService: this._stitchingService,
      url: this._url,
      dom: this._dom.getHashAsObject(),
      resources,
    }

    if (this._renderId) {
      object.renderId = this._renderId
    }

    if (this._agentId) {
      object.agentId = this._agentId
    }

    if (this._browserName) {
      object.browser = {
        name: this._browserName,
      }

      if (this._platform) {
        object.browser.platform = this._platform
      }
    }

    if (this._renderInfo) {
      object.renderInfo = this._renderInfo.toJSON()
    }

    if (this._scriptHooks) {
      object.scriptHooks = this._scriptHooks
    }

    if (this._selectorsToFindRegionsFor) {
      object.selectorsToFindRegionsFor = this._selectorsToFindRegionsFor
    }

    if (this._sendDom !== undefined) {
      object.sendDom = this._sendDom
    }

    return object
  }

  /**
   * @override
   */
  toString() {
    return `RenderRequest { ${JSON.stringify(this)} }`
  }
}

exports.RenderRequest = RenderRequest
